storage/storage_image.go

```diff
@@ -0,0 +1,411 @@
+	}
+	img, err := transport.mall.GetImage(ref)
+	if err != nil {
+		return nil
+	}
+	image := storageImage{
+		transport:      transport,
+		imageRef:       imageRef,
+		Tag:            tag,
+		Created:        time.Now(),
+		ID:             img.ID,
+		BlobList:       []string{},
+		Layers:         make(map[string]string),
+		BlobData:       make(map[string][]byte),
+		BlobDigestData: make(map[string]string),
+	}
+	image.loadMetadata()
+	return &image
+}
+
+// newImageDestination sets us up to write a new image.
+func newImageDestination(transport *storageTransport, imageRef *storageReference) *storageImage {
+	// We set the ID if the reference we got looked like one, since we take
+	// that as an indication that it's going to end up with the same
+	// contents as an image we already have.  If the reference looks more
+	// like a name, we don't know yet if it'll be exactly the same as, or
+	// an updated version of, an image we might already have, so we have to
+	// err on the side of caution and create a new image which will be
+	// assigned the name as its tag.
+	image := storageImage{
+		transport:      transport,
+		imageRef:       imageRef,
+		Tag:            imageRef.reference,
+		Created:        time.Now(),
+		ID:             imageRef.id,
+		BlobList:       []string{},
+		Layers:         make(map[string]string),
+		BlobData:       make(map[string][]byte),
+		BlobDigestData: make(map[string]string),
+	}
+	return &image
+}
+
+func newImage(transport *storageTransport, imageRef *storageReference) *storageImage {
+	return newImageSource(transport, imageRef)
+}
+
+func (s *storageImage) loadMetadata() {
+	if s.ID != "" {
+		image, err := s.transport.mall.GetImage(s.ID)
+		if image != nil && err == nil {
+			json.Unmarshal([]byte(image.Metadata), s)
+		}
+	}
+}
+
+func (s *storageImage) saveMetadata() {
+	if s.ID != "" {
+		if metadata, err := json.Marshal(s); len(metadata) != 0 && err == nil {
+			istore, err := s.transport.mall.GetImageStore()
+			if istore != nil && err == nil {
+				istore.SetMetadata(s.ID, string(metadata))
+			}
+		}
+	}
+}
+
+func (s *storageImage) Reference() types.ImageReference {
+	s.ID = s.imageRef.ID()
+	return s.imageRef
+}
+
+// PutBlob is used to both store filesystem layers and binary data that is part
+// of the image.
+func (s *storageImage) PutBlob(digest string, stream io.Reader) error {
+	// Try to read an initial snippet of the blob.
+	header := make([]byte, 10240)
+	n, err := stream.Read(header)
+	if err != nil && err != io.EOF {
+		return err
+	}
+	// Set up to read the whole blob (the initial snippet, plus the rest)
+	// while digesting it with sha256.
+	hasher := sha256.New()
+	defragmented := io.MultiReader(bytes.NewBuffer(header[:n]), stream)
+	multi := io.TeeReader(defragmented, hasher)
+	if (n > 0) && archive.IsArchive(header[:n]) {
+		// It's a filesystem layer.  If it's not the first one in the
+		// image, we assume that the most recently added layer is its
+		// parent.
```

Comment:
Sadly, looking at the current `genericImage.BlobDigests`, the order is not currently preserved on input. That can change of course, and I think `docker-daemon:` will need this to change as well.

---

For these kinds of API changes, it would be very convenient to have both the definition of the interfaces, the transports implementing the interfaces, and the core of `skopeo copy`, the primary user, in a single repo — then we could change the `ImageSource`/`ImageDestination` interfaces without breaking anything else and without needing to keep a `skopeo` branch+PR in sync.

https://github.com/containers/image/pull/56 + https://github.com/projectatomic/skopeo/pull/175 does that, so if this PR ended up changing the types, perhaps it would be practical to base it on that?

(Warning: #56 is being rebased as other PRs are merged; but I don't expect it to substantially change.)

@runcom Note that #56 currently depends on #52 , primarily for the `types.SystemContext` API changes; I could split the API changes from #52 so that #52 is only the lookaside implementation, which would allow merging the API changes + `copy.go` before reviewing the lookaside—if that helped you in any way.